jsrc/conversions.cpp

```diff
@@ -0,0 +1,637 @@
+        auto const q = jround(p);
+        if (!(p == q || fuzzy_equal(p, q, fuzz))) {
+            return std::nullopt;  // must equal int, possibly out of range
+        }
+        // out-of-range values don't convert, handle separately
+        if (p < static_cast<D> IMIN) {
+            return value_if(p >= IMIN * (1 + fuzz), IMIN);
+        }  // if tolerantly < IMIN, error; else take IMIN
+        else if (p >= FLIMAX) {
+            return value_if(p <= -static_cast<D> IMIN * (1 + fuzz), IMAX);
+        }  // if tolerantly > IMAX, error; else take IMAX
+        return q;
+    });
+}
+
+template <>
+[[nodiscard]] auto
+convert<Z, D>(J jt, array w, void *yv, D fuzz) -> bool {
+    if (fuzz != 0.0) {
+        return convert<Z, D>(jt, w, yv, [&](auto const &v) {
+            auto const d = std::abs(v.im);
+            return value_if(d != inf && d <= fuzz * std::abs(v.re), v.re);
+        });
+    }
+    return convert<Z, D>(jt, w, yv, [](auto v) { return value_if(!v.im, v.re); });
+}
+
+template <>
+[[nodiscard]] auto
+convert<bool, X>(J jt, array w, void *yv) -> bool {
+    return convert<B, X>(jt,
+                         w,
+                         yv,
+                         [=](auto v) {
+                             int64_t u[] = {v};
+                             return jtvec(jt, INT, 1L, u);
+                         }) &&
+           !jt->jerr;
+}
+
+template <typename T>
+static auto
+inplace_negate(T *u, int64_t n) {
+    std::transform(u, u + n, u, [](auto v) { return -v; });
+}
+
+template <>
+[[nodiscard]] auto
+convert<I, X>(J jt, array w, void *yv) -> bool {
+    I u[XIDIG];
+    auto const convert_one = [&](auto c) {
+        auto const b   = c == IMIN;
+        auto d         = b ? -(1 + c) : std::abs(c);
+        int64_t length = 0;
+        for (int64_t i = 0; i < XIDIG; ++i) {
+            u[i] = d % XBASE;
+            d    = d / XBASE;
+            if (u[i]) length = i;
+        }
+        ++length;
+        *u += b;
+        if (0 > c) { inplace_negate(u, XIDIG); }
+        return jtvec(jt, INT, length, u);
+    };
+    return convert<I, X>(jt, w, yv, convert_one) && !jt->jerr;
+}
+
+static auto
+jtxd1(J jt, D p, I mode) -> X {
+    PROLOG(0052);
+    D e = jttfloor(jt, p);
+    switch (mode) {
+        case XMFLR: p = e; break;
+        case XMCEIL: p = ceil(p); break;
+        case XMEXACT:
+            ASSERT(TEQ(p, e), EVDOMAIN);
+            p = e;
+            break;
+        case XMEXMT:
+            if (!TEQ(p, e)) { return jtvec(jt, INT, 0L, &iotavec[-IOTAVECBEGIN]); }
+    }
+    if (p == inf) { return jtvci(jt, XPINF); }
+    if (p == -inf) { return jtvci(jt, XNINF); }
+    array t = make_array<int64_t, copy_shape_0>(jt, 30, 1);
+    if (!t) return 0;
+    auto *u   = pointer_to_values<int64_t>(t);
+    int64_t m = 0;
+    auto d    = std::abs(p);
+    while (0 < d) {
+        auto const q = floor(d / XBASE);
+        auto const r = d - q * XBASE;
```

Comment:
Note, `divmod` doesn't exist yet - can you pretty please wr
```suggestion
        auto const [q, r] = std::div(d, XBASE); // from cstdlib
```

Reply:
The code worked just fine with `std::div()`